### PR TITLE
Customer Search: Implement UI

### PR DESCRIPTION
**Pull request by iamgabrielma**
Part of #7741
Branched from https://github.com/woocommerce/woocommerce-ios/pull/7848 , that PR needs to be merged first.

### Description
This PR implements the basic UI for searching Customers. The scope of this PR is limited to display a search view when we tap on the search icon, display Customer results matching our keyword, and pass the Customer data back to our `EditOrderAddressForm` view when we tap in any of the Customers. Filling the data to corresponding fields will be done in a future PR.

### Changes
- Adds a `OrderCustomerListView`, which is a SwiftUI wrapper for the existing `SearchViewController`.
- Adds a `CustomerSearchUICommand` class, which is the implementation of the existing `SearchUICommand` for Customer search.

### Testing instructions
1. Go to Orders > + > Create new Order > + Add Customer details > Tap on the magnifying glass icon:

3. Search for a Customer name, and tap on it:

4. Check if details are printed in the Xcode console. This will display steps 1 to 4, which represents the Customer data flow to `EditOrderAddressForm`, which we'll use to fill the Order with Customer details in a future PR.

```
1 - Customer tapped
2 - Customer ID: 5 - Name: Hello)
3 - Customer Callback. Fill Order data with Customer details
4 - Customer ID: 5 - Name: Hello)
```

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/CustomerSearchUICommand.swift [hunk 1]
```diff
@@ -0,0 +1,80 @@
+import Foundation
+import Yosemite
+
+/// Implementation of `SearchUICommand` for Customer search.
+///
+final class CustomerSearchUICommand: SearchUICommand {
+
+    typealias Model = Customer
+    typealias CellViewModel = TitleAndSubtitleAndStatusTableViewCell.ViewModel
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/CustomerSearchUICommand.swift:9]: Happy to get a different recommendation for the CellView we want to use here, at the moment I'm using `TitleAndSubtitleAndStatusTableViewCell` just for convenience, as must conform to `SearchResultCell` there are several options to choose from without having to create a new one (or I could create a new one as well!)

**Comment by iamgabrielma** [comment]
@rachelmcr @ThomazFB @Ecarrion please leave a comment if any of you plan to review the PR today, otherwise I'll review it with @joshheald during our pairing session. Thanks!

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/CustomerSearchUICommand.swift [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    func didSelectSearchResult(model: Customer, from viewController: UIViewController, reloadData: () -> Void, updateActionButton: () -> Void) {
+        // Not implemented yet
+        print("1 - Customer tapped")
+        print("2 - Customer ID: \(model.customerID) - Name: \(model.firstName ?? ""))")
+        // Customer data will go up to EditOrderAddressForm, via OrderCustomerListView completion handler
+        onDidSelectSearchResult(model)
+    }
+
+    func searchResultsPredicate(keyword: String) -> NSPredicate? {
+        return NSPredicate(format: "ANY searchResults.keyword = %@", keyword)
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/CustomerSearchUICommand.swift]: Needs to add siteID! We'll add it in a future PR/commit as this is currently behind a feature flag.

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/CustomerSearchUICommand.swift (hunk 1 above)
Comment (joshheald) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/CustomerSearchUICommand.swift:9]: Given that you're going to have the avatar in there too, you might want to try adding `SearchResultCell` conformance to `LeftImageTitleSubtitleTableViewCell`?

The only thing it won't do is any attributed text in the subtitle for an underlined email as on Android... but I don't personally think that should be a blocker.

File: WooCommerce/WooCommerce.xcodeproj/project.pbxproj [hunk 3]
```diff
@@ -11965,7 +11973,7 @@
 		};
 		4598298028574688003A9AFE /* Inject */ = {
 			isa = XCSwiftPackageProductDependency;
-			package = 4598297F28574688003A9AFE /* XCRemoteSwiftPackageReference "Inject.git" */;
+			package = 4598297F28574688003A9AFE /* XCRemoteSwiftPackageReference "Inject" */;
```
Comment (joshheald) [inline WooCommerce/WooCommerce.xcodeproj/project.pbxproj]: Has Inject changed for any particular reason? What happens if you revert these changes?

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/CustomerSearchUICommand.swift [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+    init(siteID: Int64, onDidSelectSearchResult: @escaping ((Customer) -> Void)) {
+        self.siteID = siteID
+        self.onDidSelectSearchResult = onDidSelectSearchResult
+    }
+
+    func createResultsController() -> ResultsController<StorageCustomer> {
+        let storageManager = ServiceLocator.storageManager
+        let descriptor = NSSortDescriptor(keyPath: \StorageCustomer.customerID, ascending: false)
+        return ResultsController<StorageCustomer>(storageManager: storageManager, sortedBy: [descriptor])
```
Comment (joshheald) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/CustomerSearchUICommand.swift]: This will display `StorageCustomers` from Core Data for any site, not just the current one, which could lead to confusion for some merchants with multiple stores.

If you add a predicate specifying the siteID, that should resolve the issue. No need to block this PR since it's behind a feature flag.

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/CustomerSearchUICommand.swift [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    func createStarterViewController() -> UIViewController? {
+        nil
+    }
+
+    func createCellViewModel(model: Customer) -> TitleAndSubtitleAndStatusTableViewCell.ViewModel {
+        return CellViewModel(
+            id: "\(model.customerID)",
+            title: "\(model.firstName ?? "") \(model.lastName ?? "")",
```
Comment (joshheald) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/CustomerSearchUICommand.swift:43]: This... might be the right thing to do given our data source, but names are complicated, and simply concatenating them isn't correct in all locales.

[See `PersonNameComponentsFormatter` docs](https://developer.apple.com/documentation/foundation/personnamecomponentsformatter#) for more info. There's not _necessarily_ anything to do here (as I don't know whether we have enough in our data source to correctly make a formatter) but I thought it would be useful for you to know about if you don't already.

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/CustomerSearchUICommand.swift [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+            subtitle: model.email,
+            accessibilityLabel: "",
```
Comment (joshheald) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/CustomerSearchUICommand.swift:45]: Accessibility Label is required here... see what it does in VoiceOver without it: it just announces:

> Button

Without any detail about the name of the user or anything else about the search result. 

If you're moving to another cell, you may not have to do anything, usually we get VoiceOver in cells for free, but in this case we are specific about what the label should be in the view model.

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/CustomerSearchUICommand.swift (hunk 2 above)
Comment (joshheald) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/CustomerSearchUICommand.swift]: Yep! Don't forget both places, this is the one for the `CustomerSearchResults` rather than the `Customer`

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/CustomerSearchUICommand.swift [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+    typealias Model = Customer
+    typealias CellViewModel = TitleAndSubtitleAndStatusTableViewCell.ViewModel
+    typealias ResultsControllerModel = StorageCustomer
+
+    var searchBarPlaceholder: String = Localization.searchBarPlaceHolder
+
+    var searchBarAccessibilityIdentifier: String = "customer-search-screen-search-field"
+
+    var cancelButtonAccessibilityIdentifier: String = "customer-search-screen-cancel-button"
```
Comment (joshheald) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/CustomerSearchUICommand.swift:16]: You probably only need to set accessibilityIdentifiers if you're actually using the search in a UI test. Identifiers are nothing to do with VoiceOver and other assistive tech, in practice, they are only used for UI tests. 

AccessibilityLabels are for VoiceOver and important to have on custom controls or complex groups.

**Review by joshheald — APPROVED** [review]
Well done. I was able to test it and get some search results showing in the app.

The comments here can all be addressed in a future PR, nothing is blocking.

Other general things I noticed, outside the scope of this PR, for your notes:

1. There are lots of network requests to `/wc/v3/customers/0`, which will always fail, these could be filtered out.
2. There's no empty state yet, just a blank white screen.
3. The search doesn't behave the same as on Android. On Android, search looks at the email field only, and on iOS, it looks at the name fields only. Apologies if I'm forgetting some context here, if Android are going to change that...

:shipit:

**Comment by joshheald** [comment]
General observation: from my test site, the feature feels a bit... frustrating, in terms of the search results it can actually return. This is just a limitation of the API AFAIK, but I have lots of customer details on different orders, where the customers checked out as guests. In the Analytics section of wc-admin, it manages to link things together for those customers, even though they're guests... it certainly would be good if we could do a bit more like that for filling address details! 

The brief here is to get to parity with Android, but if you do have thoughts or findings which could feed in to an iteration 2 of the feature, it would be good to hear them...

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/CustomerSearchUICommand.swift (hunk 2 above)
Comment (iamgabrielma) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/CustomerSearchUICommand.swift]: Definitely, I went ahead and changed it in this PR as was a quick one: a8dd77898ca4facaf06bf66d1d48acf7e692a846

File: WooCommerce/WooCommerce.xcodeproj/project.pbxproj (hunk 3 above)
Comment (iamgabrielma) [reply WooCommerce/WooCommerce.xcodeproj/project.pbxproj]: Good catch! And that's a good question: 

The only difference I'm seeing is that when I revert the changes is that I can see the iOS 15.0 minimum support warnings for YosemiteTests related to `kUTTypeJPEG` deprecation.

This most likely comes from me switching to iOS15 temporarily when working on this, as at the same time we were upgrading the app minimum version for the app and I wanted to see if my changes came with potential deprecations. Then when I switched back to iOS14, I may have missed the changes in `project.pbxproj`.

I see these changes are [part of trunk already](https://raw.githubusercontent.com/woocommerce/woocommerce-ios/trunk/WooCommerce/WooCommerce.xcodeproj/project.pbxproj), but I reverted here: da5e40177aed121e1ef301c86ac531d6025c010d as this change shouldn't go with my PR anyway 👍

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/CustomerSearchUICommand.swift (hunk 4 above)
Comment (iamgabrielma) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/CustomerSearchUICommand.swift]: Thanks! I committed the changes https://github.com/woocommerce/woocommerce-ios/commit/a8dd77898ca4facaf06bf66d1d48acf7e692a846 with this one as we went through these in our pairing as was pretty quick 👍

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/CustomerSearchUICommand.swift (hunk 5 above)
Comment (iamgabrielma) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/CustomerSearchUICommand.swift:43]: Thanks for the link, I definitely didn't know about `PersonNameComponentsFormatter`. I agree with you, concatenating strings doesn't look good to me either and I'll look for cleaner alternatives (if possible). I'll dig into this document and make any necessary improvements on another PR.

**Comment by iamgabrielma** [comment]
> The search doesn't behave the same as on Android. On Android, search looks at the email field only, and on iOS, it looks at the name fields only. Apologies if I'm forgetting some context here, if Android are going to change that...

Yes, this was one of the API limitations when taking this approach (pe5pgL-3r-p2)

> One big problem: search doesn’t search names, phones, and emails at the same time. You’ll have to use searchby and define what field you want to search.

I believe that, depending on when we plan to change the API behaviour, we could use this to our advantage and create some sort of "advanced search" with `searchBy`, which would also improve your point about the search feature being frustrating, which I agree with.

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/CustomerSearchUICommand.swift (hunk 7 above)
Comment (iamgabrielma) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/CustomerSearchUICommand.swift:16]: TIL!
